File: leetcode/1293.cpp
#include <cstdio>
#include <vector>
#include <algorithm>
#include <cmath>
#include <cstring>
#include <queue>

using namespace std;
int dist[41][41][1605];
void dfs(int& ans, int step,
	vector<vector<int>>& grid, int k, int used, int i, int j,
	int n, int m, vector<vector<bool>>& visited) {

	if (i == n-1 && j == m-1) {
		if (ans == -1 || ans > step) {
			ans = step;
		}
	}
	if (i-1 >= 0 && j >= 0 && visited[i-1][j] == 0) {
		visited[i-1][j] = 1;
		if (grid[i-1][j] == 1 && used < k) {
			if (dist[i-1][j][used+1] == -1 || dist[i-1][j][used+1] > step+1) {
				dist[i-1][j][used+1] = step+1;
				dfs(ans, step+1, grid, k, used+1, i-1, j, n, m, visited);
			}	
		} else if (grid[i-1][j] == 0) {
			if (dist[i-1][j][used] == -1 || dist[i-1][j][used] > step+1) {
				dist[i-1][j][used] = step+1;
				dfs(ans, step+1, grid, k, used, i-1, j, n, m, visited);
			}
		}
		visited[i-1][j] = 0;
	}
	if (i+1 < n && j >= 0 && visited[i+1][j] == 0) {
		visited[i+1][j] = 1;
		if (grid[i+1][j] == 1 && used < k) {
			if (dist[i+1][j][used+1] == -1 || dist[i+1][j][used+1] > step+1) {
				dist[i+1][j][used+1] = step+1;
				dfs(ans, step+1, grid, k, used+1, i+1, j, n, m, visited);
			}
			
		} else if (grid[i+1][j] == 0) {
			if (dist[i+1][j][used] == -1 || dist[i+1][j][used] > step+1) {
				dist[i+1][j][used] = step+1;
				dfs(ans, step+1, grid, k, used, i+1, j, n, m, visited);
			}
		}
		visited[i+1][j] = 0;
	}
	if (i >= 0 && j-1 >= 0 && visited[i][j-1] == 0) {
		visited[i][j-1] = 1;
		if (grid[i][j-1] == 1 && used < k) {
			if (dist[i][j-1][used+1] == -1 || dist[i][j-1][used+1] > step+1) {
				dist[i][j-1][used+1] = step+1;
				dfs(ans, step+1, grid, k, used+1, i, j-1, n, m, visited);
			}
			
		} else if (grid[i][j-1] == 0) {
			if (dist[i][j-1][used] == -1 || dist[i][j-1][used] > step+1) {
				dist[i][j-1][used] = step+1;
				dfs(ans, step+1, grid, k, used, i, j-1, n, m, visited);
			}
			
		}
		visited[i][j-1] = 0;
	}
	if (i >= 0 && j+1 < m && visited[i][j+1] == 0) {
		visited[i][j+1] = 1;
		if (grid[i][j+1] == 1 && used < k) {
			if (dist[i][j+1][used+1] == -1 || dist[i][j+1][used+1] > step+1) {
				dist[i][j+1][used+1] = step+1;
				dfs(ans, step+1, grid, k, used+1, i, j+1, n, m, visited);
			}
		} else if (grid[i][j+1] == 0) {
			if (dist[i][j+1][used] == -1 || dist[i][j+1][used] > step+1) {
				dist[i][j+1][used] = step+1;
				dfs(ans, step+1, grid, k, used, i, j+1, n, m, visited);
			}
		}
		visited[i][j+1] = 0;
	}
}
int solve(vector<vector<int>>& grid, int k) {
	int ans = -1;
	int n = (int)grid.size();
	int m = (int)grid[0].size();
	int step = 0;
	int used = 0;
	vector<vector<bool>> visited(n, vector<bool>(m, false));
	// vector<vector<vector<int>>> dist(n, vector<vector<int>>(m, vector<int>(k+1, 100000)));
	memset(dist, -1, sizeof(dist));
    printf("%d %d %d\n", n, m, k);
	dfs(ans, step, grid, k, used, 0, 0, n, m, visited);
    return ans;
}

struct I {
	int i;
	int j;
	int k;
	int step;
};
int solve2(vector<vector<int>>& grid, int k) {
	int ans = -1;
	int n = (int)grid.size();
	int m = (int)grid[0].size();
	vector<vector<vector<int>>> dist(n, vector<vector<int>>(m, vector<int>(k+1, 100000)));
	queue<I> q = {};
	q.push(I{0,0,0,0});
	while (!q.empty()) {
		I it = q.front();
		q.pop();
		if (it.i == n-1 && it.j == m-1) {
			if (ans == -1 || ans > it.step) {
				ans = it.step;
			}
		}
		vector<pair<int, int>> ds = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
		for (auto x: ds) {
			if (it.i + x.first >= 0 && it.i + x.first < n && it.j + x.second >= 0 && it.j + x.second < m) {
				int a = it.i+x.first;
				int b = it.j+x.second;
				if (grid[a][b] == 1) {
					if (it.k+1 <= k && dist[a][b][it.k+1] > it.step+1) {
						dist[a][b][it.k+1] = it.step+1;
						q.push(I{a,b,it.k+1,it.step+1});
					}
				} else {
					if (it.k <= k && dist[a][b][it.k] > it.step+1) {
						dist[a][b][it.k] = it.step+1;
						q.push(I{a,b,it.k,it.step+1});
					}
				}	
			}
		}
	}
	return ans;
}
int main() {
	printf("%d\n", solve());
	return 0;
}

File: leetcode/old/1293.go
package main

import (
	"fmt"
	"math/rand"
)

func dfs(ans int, step int, grid [][]int, k int, used int, i int, j int, n int, m int, visited [][]int) int {
	if i == n-1 && j == m-1 {
		if ans == -1 || ans > step {
			ans = step
		}
		return ans
	}
	if i-1 >= 0 && j >= 0 && visited[i-1][j] == 0 {
		visited[i-1][j] = 1
		if grid[i-1][j] == 1 && used < k {
			ans = dfs(ans, step+1, grid, k, used+1, i-1, j, n, m, visited)
		} else if grid[i-1][j] == 0 {
			ans = dfs(ans, step+1, grid, k, used, i-1, j, n, m, visited)
		}
		visited[i-1][j] = 0
	}
	if i+1 < n && j >= 0 && visited[i+1][j] == 0 {
		visited[i+1][j] = 1
		if grid[i+1][j] == 1 && used < k {
			ans = dfs(ans, step+1, grid, k, used+1, i+1, j, n, m, visited)
		} else if grid[i+1][j] == 0 {
			ans = dfs(ans, step+1, grid, k, used, i+1, j, n, m, visited)
		}
		visited[i+1][j] = 0
	}
	if i >= 0 && j-1 >= 0 && visited[i][j-1] == 0 {
		visited[i][j-1] = 1
		if grid[i][j-1] == 1 && used < k {
			ans = dfs(ans, step+1, grid, k, used+1, i, j-1, n, m, visited)
		} else if grid[i][j-1] == 0 {
			ans = dfs(ans, step+1, grid, k, used, i, j-1, n, m, visited)
		}
		visited[i][j-1] = 0
	}
	if i >= 0 && j+1 < m && visited[i][j+1] == 0 {
		visited[i][j+1] = 1
		if grid[i][j+1] == 1 && used < k {
			ans = dfs(ans, step+1, grid, k, used+1, i, j+1, n, m, visited)
		} else if grid[i][j+1] == 0 {
			ans = dfs(ans, step+1, grid, k, used, i, j+1, n, m, visited)
		}
		visited[i][j+1] = 0
	}
	return ans
}

func shortestPath(grid [][]int, k int) int {
	ans := -1
	used := 0
	n := len(grid)
	m := len(grid[0])
	visited := make([][]int, n)
	for i := 0; i < n; i++ {
		visited[i] = make([]int, m)
	}
	return dfs(ans, 0, grid, k, used, 0, 0, n, m, visited)
}
func main() {
	// fmt.Println(shortestPath([][]int{
	// 	{0, 0, 0},
	// 	{1, 1, 0},
	// 	{0, 0, 0},
	// 	{0, 1, 1},
	// 	{0, 0, 0},
	// }, 1))
	rand.Seed(86)
	grid := make([][]int, 40)
	for i := 0; i < 40; i++ {
		grid[i] = make([]int, 40)
		for j := 0; j < 40; j++ {
			t := rand.Intn(2)
			if t > 1 {
				fmt.Println("dsadas")
			}
			grid[i][j] = t
		}
	}
	grid[0][0] = 0
	grid[39][39] = 0
	k := rand.Intn(1601)
	if k == 0 {
		k++
	}
	fmt.Println("GOGOGO")
	fmt.Println(shortestPath(grid, k))
}
